crates/hotfix/tests/common/mock_counterparty.rs

```diff
@@ -1,67 +1,74 @@
-use hotfix::message::{FixMessage, RawFixMessage};
+use hotfix::message::FixMessage;
 use hotfix::session::SessionRef;
 use hotfix::transport::FixConnection;
 use hotfix::transport::reader::ReaderRef;
 use hotfix::transport::writer::{WriterMessage, WriterRef};
-use std::sync::Arc;
-use std::time::{Duration, Instant};
-use tokio::sync::mpsc::Receiver;
-use tokio::sync::{Mutex, mpsc, oneshot};
+use std::time::Duration;
+use tokio::sync::{mpsc, oneshot};
+
+const DEFAULT_TIMEOUT: Duration = Duration::from_secs(1);
 
 pub struct MockCounterparty {
-    received_messages: Arc<Mutex<Vec<RawFixMessage>>>,
+    // Receiver-End of the channel
+    receiver: mpsc::UnboundedReceiver<WriterMessage>,
+    // History of Received messages from the client
+    messages: Vec<WriterMessage>,
     _connection: FixConnection,
     _dc_sender: oneshot::Sender<()>,
 }
 
-type MessageStore = Arc<Mutex<Vec<RawFixMessage>>>;
-
 impl MockCounterparty {
     pub async fn start(session_ref: SessionRef<impl FixMessage>) -> Self {
-        let (writer_ref, received_messages) = Self::spawn_writer();
+        let (writer_ref, receiver) = Self::spawn_writer();
         let (reader_ref, dc_sender) = Self::create_reader();
         let connection = FixConnection::new(writer_ref, reader_ref);
 
         session_ref.register_writer(connection.get_writer()).await;
 
         Self {
-            received_messages,
+            receiver,
+            messages: vec![],
             _connection: connection,
             _dc_sender: dc_sender,
         }
     }
 
-    pub async fn assert_message_count(&self, expected_count: usize, timeout_secs: f32) {
-        let timeout_duration = Duration::from_secs_f32(timeout_secs);
-        let start_time = Instant::now();
-
-        loop {
-            {
-                let messages = self.received_messages.lock().await;
-                if messages.len() >= expected_count {
-                    return;
-                }
-            }
+    /// Listen to the next message on the channel
+    pub async fn get_next(&mut self, timeout: Option<Duration>) -> Option<&WriterMessage> {
```

Comment:
I'm not sure I'd make the return value an `Option` here. You might as well fail if it's `None`, no?

Especially with your implementation, where you send the disconnect message back as well, you're pretty much guaranteed to get the disconnect message before a `None`.